File: LinkedList5/MediumOfLL/AddTwoNumbers.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class AddTwoNumbers:

    def addTwoNumbers(self, l1, l2): 
        
        dummyNode = ListNode(None)
        dummy = dummyNode

        carry = 0

        while l1 and l2:
            sum = l1.val + l2.val + carry
            dummy.next = ListNode(sum % 10)
            carry = sum // 10;

            l1 = l1.next
            l2 = l2.next
            dummy = dummy.next
        
        while l1:
            sum = l1.val + carry
            dummy.next = ListNode(sum%10)
            carry = sum//10;

            l1 = l1.next
            dummy = dummy.next

        while l2:
            sum = l2.val + carry
            dummy.next = ListNode(sum%10)
            carry = sum//10;

            l2 = l2.next
            dummy = dummy.next
        
        while carry >0:
            dummy.next = ListNode(carry%10)
            carry//=10


        return dummyNode.next

node1 = ListNode(2)
node1.next = ListNode(4)
node1.next.next = ListNode(3)

node2 = ListNode(5)
node2.next = ListNode(6)
node2.next.next = ListNode(4)

obj = AddTwoNumbers()
head = obj.addTwoNumbers(node1, node2);


while head:
    print(head.val)
    head = head.next

File: LinkedList5/addTwoNumbers.java
// https://leetcode.com/problems/add-two-numbers-ii/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if(l1 == null) return l2;
        if(l2 == null) return l1;

        int[] a = new int[100];
        int[] b = new int[100];

        int i=0, j=0;
        ListNode c1 = l1, c2 = l2;

        while(c1 != null){
            a[i++] = c1.val;
            c1 = c1.next;
        }
        
        while(c2!= null) {
            b[j++] = c2.val;
            c2 = c2.next;
        }

        i = i - 1;
        j = j - 1;

        ListNode dummy = new ListNode();
        ListNode current = dummy;

        int carry = 0;
        while( i >= 0 && j>=0){
            int val = a[i] + b[j] + carry;
            
            ListNode newNode = new ListNode(val%10, current.next);
            current.next = newNode;

            carry = val/10;
            i-=1;
            j-=1;
        }

        while( i >= 0){
            int val = a[i] + carry;
            
            ListNode newNode = new ListNode(val%10, current.next);
            current.next = newNode;

            carry = val/10;
            i-=1;
        }

        while( j>=0){
            int val = b[j] + carry;
            
            ListNode newNode = new ListNode(val%10, current.next);
            current.next = newNode;

            carry = val/10;
            j-=1;
        }

        while(carry != 0){
            ListNode newNode = new ListNode(carry%10, current.next);
            current.next = newNode;

            carry/=10;
        }

        
        return dummy.next;
    }
}



// ----------------------------------------------------------------------------

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if(l1 == null) return l2;
        if(l2 == null) return l1;

        int[] a = new int[100];
        int[] b = new int[100];

        int i=0, j=0;
        ListNode c1 = l1, c2 = l2;

        while(c1 != null){
            a[i++] = c1.val;
            c1 = c1.next;
        }
        
        while(c2!= null) {
            b[j++] = c2.val;
            c2 = c2.next;
        }

        i = i - 1;
        j = j - 1;

        ListNode dummy = new ListNode();
        ListNode current = dummy;

        int carry = 0;
        while( i >= 0 || j>=0 || carry != 0){
            int sum = carry;
            if(i >= 0) sum += a[i];
            if(j >= 0) sum += b[j] ;
            
            ListNode newNode = new ListNode(sum%10, current.next);
            current.next = newNode;

            carry = sum/10;
            i-=1;
            j-=1;
        }
        
        return dummy.next;
    }
}